Give std/var correction overloads proper defaults

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #55680
* #55679
* #87118
* __->__ #56398

The correction overloads defaults were left off for forward
compatibility reasons, but this FC window expired well over a year ago
at this point.

Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @jansel @lezcano @fdrocha

peterbell10 committed:
Give std/var correction overloads proper defaults

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

peterbell10 commented:
@mruberry PTAL, forward-compat period should be over now.

mruberry commented on torch/_tensor_docs.py:
```diff
@@ -3570,7 +3570,12 @@ def callable(a, b) -> number
+
+See :func:`torch.std`
+
+.. function:: std(dim, unbiased=True, keepdim=False) -> Tensor
```
Let's just remove these legacy signatures from the docs

mruberry replied:
Can we remove these or were you thinking we should keep them?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8461,7 +8461,48 @@ def merge_dicts(*dicts):
+            [1.2204],
+            [0.9087]])
+
+.. function:: std(input, dim, unbiased, keepdim=False, *, out=None) -> Tensor
```
Let's kill these definitions, too

mruberry commented on torch/_torch_docs.py:
```diff
@@ -9724,7 +9808,48 @@ def merge_dicts(*dicts):
-var(input, dim, unbiased, keepdim=False, *, out=None) -> Tensor
+var(input, dim=None, *, correction=1, keepdim=False, out=None) -> Tensor
+
+Calcualtes the variance over the dimensions specified by :attr:`dim`. :attr:`dim`
```
"Calcualtes" -> "Calculates"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8461,7 +8461,48 @@ def merge_dicts(*dicts):
+""" + r"""
+
+{keepdim_details}
+
```
Let's add a note here (and to the other functions about the historic behavior):

"In PyTorch 1.9 and earlier _std_ accepted an _unbiased_ argument. If _unbiased_ was True (the default) it implied `correction=1` (the current default). If _unbiased_ was False then it implied `correction=0`. The new _correction_ parameter generalizes the deprecated _unbiased_ parameter."

How does that sound?

peterbell10 replied:
Tweaked the wording slightly to:
>   In PyTorch 1.8 and earlier `std` accepted an `unbiased` argument which was a less general form of the `correction` parameter. If `unbiased` was ``True`` it implied ``correction=1`` (this was the default). If `unbiased` was ``False`` then it implied ``correction=0``.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8461,7 +8461,48 @@ def merge_dicts(*dicts):
+
+Keyword args:
+    correction (int): difference between the sample size and sample degrees of freedom.
+                      Defaults to Bessel's correction, ``correction = 1``.
```
Is it worth mentioning that when correction = 1 this is an unbiased estimator of the standard deviation?

peterbell10 replied:
I've added a link to the wiki for Bessel's correction. IMO It's a bit more complicated than just being unbiased, it would have to be more like 
> assuming the input is a sample from a normally-distributed population, this gives an unbiased estimator of the population standard deviation.

which is probably too wordy for an argument description.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -9724,7 +9808,48 @@ def merge_dicts(*dicts):
+
+Keyword args:
+    correction (int): difference between the sample size and sample degrees of freedom.
+                      Defaults to Bessel's correction, ``correction = 1``.
```
Is it worth noting that when correction = 1 this is an unbiased estimator of the population variance?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2673,7 +2673,7 @@ def sample_inputs_std_var(op_info, device, dtype, requires_grad, **kwargs):
 
         SampleInput(tensor_nd, kwargs=dict(dim=(1,), correction=S // 2)),
-        SampleInput(tensor_nd, kwargs=dict(dim=None, correction=0, keepdim=True)),
+        SampleInput(tensor_nd, kwargs=dict(correction=0, keepdim=True)),
```
Let's change the above "unbiased" uses to use "correction" and add a SampleInput with a correction factor that is not 0 or 1

peterbell10 replied:
>Let's change the above "unbiased" uses to use "correction" 

The unbiased overloads are needed for backwards-compatibility so IMO they should be tested.

>add a SampleInput with a correction factor that is not 0 or 1

The line above this has `correction=S//2` which is 2.

mruberry replied:
Cool

mruberry reviewed (commented):
Thanks for the ping, @peterbell10!

This looks really good. I made a few comments inline. Basically I want to hide the old parameter, add a note explaining that it existed and how people should update to the new functionality, and consider elaborating on the mathematical impact of the correction factor.

Looking forward to hearing your thoughts.

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

peterbell10 commented:
@mruberry PTAL

mruberry commented on test/test_namedtensor.py:
```diff
@@ -1241,26 +1245,26 @@ def kthvalue_wrapper(tensor, *args, **kwargs):
-            Case(kthvalue_wrapper, False, False, True, True, values_and_indices),
-            Case(torch.median, True, False, True, True, values_and_indices),
-            Case(torch.nanmedian, True, False, True, True, values_and_indices),
+            Case(torch.sum, False, True, True, True, True, None),
```
fyi @heitorschueroff; it'll be nice to revise this test so we don't deal with all these unnamed tuples

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8448,81 +8448,110 @@ def merge_dicts(*dicts):
+the :attr:`correction`.
+""" + r"""
+
+Note:
```
Nice note.

mruberry replied:
Should this be converted to a version changed directive now? Also, it would be PyTorch 1.13?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -9711,80 +9740,109 @@ def merge_dicts(*dicts):
-
-.. function:: var_mean(input, unbiased) -> (Tensor, Tensor)
-   :noindex:
+    A tuple (var, mean) containing the variance and mean. Here 2
```
I think the "Here 2" is extraneous?

peterbell10 replied:
Thanks, that was left over from testing that the docs were rendering properly. Removed.

mruberry reviewed (commented):
Looks really good, @peterbell10, just a few small comments inline

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

mruberry commented:
Everything looks good, @peterbell10. Last question is if this should remove the documentation of the `unbiased` signatures in torch/_tensor_docs.py? I think so, right?

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"




[ghstack-poisoned]

peterbell10 commented:
True, it should match the function docs. Fixed.

mruberry reviewed (approved):
Awesome! Thanks @peterbell10!

mruberry commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -79,6 +79,10 @@
     ("aten::divide", datetime.date(2021, 4, 28)),
     ("aten::_var", datetime.date(2021, 5, 28)),
     ("aten::_std", datetime.date(2021, 5, 28)),
+    ("aten::std", datetime.date(2021, 5, 28)),
```
Last issue: these entries need to be updated since the old dates expired:

```
Jul 07 19:50:17 Broken ops: [
Jul 07 19:50:17 	aten::std_mean(Tensor self, bool unbiased=True) -> (Tensor, Tensor)
Jul 07 19:50:17 	aten::std_mean.dim(Tensor self, int[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor, Tensor)
Jul 07 19:50:17 	aten::std_mean.names_dim(Tensor self, str[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor, Tensor)
Jul 07 19:50:17 	aten::std(Tensor self, bool unbiased=True) -> (Tensor)
Jul 07 19:50:17 	aten::std.dim(Tensor self, int[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor)
Jul 07 19:50:17 	aten::std.names_dim(Tensor self, str[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor)
Jul 07 19:50:17 	aten::std.names_out(Tensor self, str[1] dim, bool unbiased=True, bool keepdim=False, *, Tensor(a!) out) -> (Tensor(a!))
Jul 07 19:50:17 	aten::std.out(Tensor self, int[1] dim, bool unbiased=True, bool keepdim=False, *, Tensor(a!) out) -> (Tensor(a!))
Jul 07 19:50:17 	aten::var(Tensor self, bool unbiased=True) -> (Tensor)
Jul 07 19:50:17 	aten::var.dim(Tensor self, int[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor)
Jul 07 19:50:17 	aten::var.names_dim(Tensor self, str[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor)
Jul 07 19:50:17 	aten::var.names_out(Tensor self, str[1] dim, bool unbiased=True, bool keepdim=False, *, Tensor(a!) out) -> (Tensor(a!))
Jul 07 19:50:17 	aten::var.out(Tensor self, int[1] dim, bool unbiased=True, bool keepdim=False, *, Tensor(a!) out) -> (Tensor(a!))
Jul 07 19:50:17 	aten::var_mean(Tensor self, bool unbiased=True) -> (Tensor, Tensor)
Jul 07 19:50:17 	aten::var_mean.dim(Tensor self, int[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor, Tensor)
Jul 07 19:50:17 	aten::var_mean.names_dim(Tensor self, str[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor, Tensor)
```

peterbell10 replied:
Ah, right. I've updated the date and rebased on a newer master.

mruberry commented:
Sorry @peterbell10, one last update. I'll start testing this internally now

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29625593).

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

mruberry commented:
I think the date has to be valid:

```
Jul 09 01:15:08   File "check_backward_compatibility.py", line 82, in <module>
Jul 09 01:15:08     ("aten::std", datetime.date(2021, 8, 0)),
Jul 09 01:15:08 ValueError: day is out of range for month
```

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 commented:
Dates are fixed now and the backward compatibility check is passing.

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29625593).

peterbell10 commented:
@mruberry any update on this PR? Were there issue with FC/BC that need to be fixed?

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

mruberry commented:
> @mruberry any update on this PR? Were there issue with FC/BC that need to be fixed?

No update yet on this PR yet, but I'm working with our compiler and mobile teams now to review this and other BC/FC issues. I hope we'll have something sorted soonish.

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 committed:
Rebase on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

lezcano commented:
What's happened to the compiler @mruberry?

peterbell10 committed:
Add test for correction=None on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


The correction overloads defaults were left off for forward
compatibility reasons, but this FC window expired well over a year ago
at this point.

Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


The correction overloads defaults were left off for forward
compatibility reasons, but this FC window expired well over a year ago
at this point.

Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 committed:
Rebase and fix merge conflicts on "Give std/var correction overloads proper defaults"


The correction overloads defaults were left off for forward
compatibility reasons, but this FC window expired well over a year ago
at this point.

Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


The correction overloads defaults were left off for forward
compatibility reasons, but this FC window expired well over a year ago
at this point.

Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

[ghstack-poisoned]

peterbell10 commented:
@mruberry PTAL, this shouldn't have any BC implications any more.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -10649,45 +10667,58 @@ def merge_dicts(*dicts):
+"""
+    + r"""
+
+Note:
```
Same question about the note here

mruberry reviewed (approved):
Looks good -- see suggestion to modernize the documentation slightly. It's unfortunately this causes some `out` tests to fail now -- we should probably get that fixed

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


The correction overloads defaults were left off for forward
compatibility reasons, but this FC window expired well over a year ago
at this point.

Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

cc jansel lezcano fdrocha

[ghstack-poisoned]

peterbell10 commented:
The `test_out` issue is that the `var(Tensor self, bool unbiased=True)` overload doesn't support an out argument, but the other overloads do. I don't really want to add out support to the deprecated overload, but I suppose we could just make that overload its own opinfo variant?

peterbell10 committed:
Rebase and fix merge conflicts on "Give std/var correction overloads proper defaults"


The correction overloads defaults were left off for forward
compatibility reasons, but this FC window expired well over a year ago
at this point.

Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

cc jansel lezcano fdrocha

[ghstack-poisoned]

peterbell10 commented:
By the way, whats the best way to deal with the "Meta Internal-Only Changes Check". I assume that's no longer relevant so can the PR  be unimported, or would force-merging work be safe, or does this have to be landed internally now.

mruberry commented:
> The `test_out` issue is that the `var(Tensor self, bool unbiased=True)` overload doesn't support an out argument, but the other overloads do. I don't really want to add out support to the deprecated overload, but I suppose we could just make that overload its own opinfo variant?

Yes, that would work. There are other operators with multiple OpInfos for similar reasons.

mruberry commented:
> By the way, whats the best way to deal with the "Meta Internal-Only Changes Check". I assume that's no longer relevant so can the PR be unimported, would force-merging be safe, or does this have to be landed internally now?



Good question. @malfet @seemethere, what do you recommend?

peterbell10 committed:
Split std/var opinfos into two on "Give std/var correction overloads proper defaults"


The correction overloads defaults were left off for forward
compatibility reasons, but this FC window expired well over a year ago
at this point.

Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

cc jansel lezcano fdrocha mlazos soumith voznesenskym yanboliang

[ghstack-poisoned]

peterbell10 commented:
@malfet @seemethere ping

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29625593).

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


The correction overloads defaults were left off for forward
compatibility reasons, but this FC window expired well over a year ago
at this point.

Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx jansel lezcano fdrocha

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


The correction overloads defaults were left off for forward
compatibility reasons, but this FC window expired well over a year ago
at this point.

Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx jansel lezcano fdrocha

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


The correction overloads defaults were left off for forward
compatibility reasons, but this FC window expired well over a year ago
at this point.

Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx jansel lezcano fdrocha

[ghstack-poisoned]

peterbell10 committed:
Update on "Give std/var correction overloads proper defaults"


The correction overloads defaults were left off for forward
compatibility reasons, but this FC window expired well over a year ago
at this point.

Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx jansel lezcano fdrocha

[ghstack-poisoned]

peterbell10 commented:
@malfet were there any mobile issues?

lezcano commented:
@malfet what's the state of this PR? Can we land it?

malfet commented:
@pytorchbot rebase

malfet commented:
Sorry for the late reply, but can you please rebase again the latest PyTorch and I'll try to import it again (as you can see it can not be imported as is)

peterbell10 committed:
Rebase on "Give std/var correction overloads proper defaults"


The correction overloads defaults were left off for forward
compatibility reasons, but this FC window expired well over a year ago
at this point.

Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx jansel lezcano fdrocha

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29625593).

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29625593).

malfet commented:
Ok, just landed internally, I wonder if it will be auto-closed by the bot now...

ngimel commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4960,10 +4960,16 @@ def sample_inputs_std_var(op_info, device, dtype, requires_grad, **kwargs):
+    yield SampleInput(tensor_nd(), correction=0, keepdim=True)
+
+
+def sample_inputs_std_var_unbiased(op_info, device, dtype, requires_grad, **kwargs):
```
why do we need unbiased variants of OpInfos? Can't we add them to the regular sample inputs function?

peterbell10 replied:
The `unbiased` variant has `supports_out=False` due to some overloads not having an equivalent `out` overload, while `correction` overloads supports the `out` argument on all signatures.
